File: The Twin Towers/challenge.py
import sys

# IMPORTANTE, NO FUNCIONA Y NO ENTIENDO POR QUE (me da el mismo output que los casos de udebug). SERA CONSULTADO HOY MISMO SI ME DAN BOLA

def LCS(tile_a, tile_b):
    '''
        Devuelve la longitud de la subsecuencia mas comun
    '''
    # Inicializo la dimension fisica de la columna y la fila, + 1 por el caso de null
    n = len(tile_a) + 1
    m = len(tile_b) + 1
    
    # Creo la matriz toda en 0 porque la vamos a usar como matriz contador
    memo = [[0 for row in range(m)] for column in range(n)]
    
    # Ahora la recorro y voy contando en la matriz
    for I in range(1, n):
        for J in range(1, m):
            if tile_a[I-1] == tile_b[J-1]:  # Si coinciden sumo 1 a la posicion actual + el diagonal para atras arriba de donde estoy parado
                memo[I][J] = memo[I-1][J-1] + 1
            else:   # Si no son iguales me quedo con el maximo entre arriba o a la izq
                memo[I][J] = max(memo[I-1][J], memo[I][J-1])
                
    return memo[n-1][m-1]

sequence = 1
while True:
    number_of_tiles = list(map(int, sys.stdin.readline().split()))
    
    if 0 in number_of_tiles:
        break
    
    tile_a = list(map(int, sys.stdin.readline().split()))
    tile_b = list(map(int, sys.stdin.readline().split()))
    
    sys.stdout.write(f"Twin Towers #{sequence}\n")
    sys.stdout.write(f"Number of Tiles : {LCS(tile_a, tile_b)}\n")
    

File: The Twin Towers/challenge.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int LCS(const vector<int>& tile_a, const vector<int>& tile_b) {
    int n = tile_a.size() + 1;
    int m = tile_b.size() + 1;

    vector<vector<int>> memo(n, vector<int>(m, 0));

    for (int i = 1; i < n; ++i) {
        for (int j = 1; j < m; ++j) {
            if (tile_a[i-1] == tile_b[j-1]) {
                memo[i][j] = memo[i-1][j-1] + 1;
            } else {
                memo[i][j] = max(memo[i-1][j], memo[i][j-1]);
            }
        }
    }

    return memo[n-1][m-1];
}

int main() {
    int sequence = 1;

    while (true) {
        int N1, N2;
        cin >> N1 >> N2;

        if (N1 == 0 && N2 == 0) {
            break;
        }

        vector<int> tile_a(N1);
        vector<int> tile_b(N2);

        for (int i = 0; i < N1; ++i) {
            cin >> tile_a[i];
        }
        for (int i = 0; i < N2; ++i) {
            cin >> tile_b[i];
        }

        int result = LCS(tile_a, tile_b);

        cout << "Twin Towers #" << sequence << endl;
        cout << "Number of Tiles : " << result << endl;
        cout << endl;

        sequence++;
    }

    return 0;
}